### PR TITLE
feat(#major); pricelib; make pricelib configurable

### PR DESCRIPTION
**Pull request by dhruv-chauhan**
**Option to configure**

{number of oracle to fetch price from, and their order of preference} based on {token addresses and/or block range}

This is done by adding an override rule for known discrepancies, and conditionally executing it:
```
class spellOverride implements OracleConfig {
  oracleCount(): number {
    return INT_ONE;
  }
  oracleOrder(): string[] {
    return [
      OracleType.CHAINLINK_FEED,
      OracleType.CURVE_CALCULATIONS,
      OracleType.SUSHI_CALCULATIONS,
      OracleType.CURVE_ROUTER,
      OracleType.UNISWAP_FORKS_ROUTER,
      OracleType.YEARN_LENS_ORACLE,
    ];
  }
}

if ([Address.fromString("0x090185f2135308bad17527004364ebcc2d37e5f6")].includes(tokenAddr)) {
    return new spellOverride();
} else if (
    [Address.fromString("0x...a"), Address.fromString("0x...b")].includes(tokenAddr) && 
    block.number.gt(100000) && block.number.lt(999999)) {
    return new someOtherOverride();
} else {
    return new DefaultOracleConfig();
}
```
Any new rules shall be maintained in both the subgraph as well as the _reference_ pricelib directory, so the same inconsistencies do not have to be handled separately. 

**Bounding prices by liquidity**

{For Uniswap-forks and Curve Router oracles}

Limits the token price by available counter-liquidity. 
If (tokenPrice*amount) > counterLiquidity, then returnedPrice is (counterLiquidity/amount).

**Averaging on k-closest reported prices**

If the user intents to query multiple oracles for the price of a token, the result is an average price on the k-closest reported values.

where, k is set to be the ceil(reportedPrices.length/2)

This helps us ignore (and spot) outlier values, by trusting the closer majority values and avoids complete reliance on a single source of truth. Also useful for low/distributed liquidity tokens, or during volatile markets.

Multiple contract calls are an overhead, but should not be the worst for smaller subgraphs, and can be used for specific tokens/block ranges for larger ones.

> PR #2190  is an implementation of thess change for stETH in lido, SPELL in arbitrum-one-bridge, and BAXA in optimism-bridge-v2 subgraphs.

### CODE REVIEW

File: subgraphs/_reference_/src/prices/common/types.ts [hunk 1]
```diff
@@ -35,22 +35,27 @@ export class CustomPriceType {
   private _usdPrice: Wrapped<BigDecimal>;
   private _decimals: Wrapped<i32>;
   private _oracleType: string;
+  private _counterLiquidity: Wrapped<BigDecimal>;
```
Comment (dmelotik) [inline subgraphs/_reference_/src/prices/common/types.ts:38]: Why not just call it `_liquidity` I am slightly confused why it is called `_counterLiquidity`

**Review by dmelotik — COMMENTED** [review]
Looks good! Few comments, also wanted to see it working on a subgraph

File: subgraphs/_reference_/src/prices/index.ts [hunk 2]
```diff
@@ -41,88 +41,92 @@ export function getUsdPricePerToken(
...
+  tokenAddress: Address,
+  tokenPrice: CustomPriceType,
+  amount: BigDecimal
+): BigDecimal {
+  const reportedPriceUSD = tokenPrice.usdPrice.times(amount);
+  const counterLiquidity = tokenPrice.counterLiquidity;
+
+  let liquidityBoundPriceUSD = reportedPriceUSD;
+  if (
+    counterLiquidity > constants.BIGDECIMAL_ZERO &&
```
[the remark this answers is not in the document]
Comment (dhruv-chauhan) [reply subgraphs/_reference_/src/prices/index.ts:105]: consider the check `liquidity > constants.BIGDECIMAL_ZERO` to be the same as if liquidity exists. 

(CustomPriceType's liquidity getter does not let me return a `BigDecimal | null` type and gives some weird graph-cli compile error, hence the return type is only `BigDecimal` with BIGDECIMAL_ZERO as the default value.)

**Review by dmelotik — APPROVED** [review]
nice work!

**Pull request merged**